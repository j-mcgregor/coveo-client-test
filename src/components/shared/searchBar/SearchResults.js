/* eslint-disable operator-linebreak */
import { get } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import { withTheme } from 'styled-components';
import S from '../../../static/styles';
import stringToKey from '../../../utils/createKey';
import Spinner from '../Spinner';
import StarRating from '../StarRating';

const SearchResults = ({ searchResults, loading, selectProduct }) => {
  const results = searchResults.map((result, i) => {
    const price = get(result, 'raw.tpprixnormal');
    const id = get(result, 'raw.tpproductid');
    const ml = get(result, 'raw.tpformat');
    const category = get(result, 'raw.tpcategorie');
    return (
      <Link
        className="list-item columns"
        to={`/products/${id}`}
        key={stringToKey(result.title, i)}
        onClick={() => selectProduct(result)}
      >
        <div className="column col-xs-6">
          <div className="columns">
            <div className="column col-12">{result.title}</div>
            <div className="column col-12">
              <small>{category}</small>
            </div>
          </div>
        </div>
        <div className="column col-xs-6 text-right">
          <div className="columns">
            <div className="column col-12">
              <StarRating percentScore={result.percentScore} />
            </div>
            <div className="column col-12">
              <small>
                {price} / {ml}
              </small>
            </div>
          </div>
        </div>
      </Link>
    );
  });

  return (
    <S.SearchResults>
      {loading ? (
        <S.SpinnerContainer>
          {' '}
          <Spinner />
        </S.SpinnerContainer>
      ) : (
        <div className="container results">{results}</div>
      )}
    </S.SearchResults>
  );
};

SearchResults.propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.shape({})),
  loading: PropTypes.bool,
  selectProduct: PropTypes.func
};

SearchResults.defaultProps = {
  searchResults: [],
  loading: false,
  selectProduct: () => {}
};

export default withTheme(SearchResults);
