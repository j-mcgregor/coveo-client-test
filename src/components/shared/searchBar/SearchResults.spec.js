import React from 'react';
import { shallow, mount } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';
import SearchResults from './SearchResults';

describe('SearchResults', () => {
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      searchResults: [],
      selectProduct: jest.fn(),
      loading: false
    };
  });

  it('should render and the items if present', () => {
    const props = {
      ...defaultProps,
      searchResults: [
        {
          title: 'Some title',
          raw: {
            tpprixnormal: 1.5,
            tpproductid: '123',
            tpformat: '330ml',
            tpcategorie: 'beer'
          },
          percentScore: 12.34
        }
      ]
    };
    const wrapper = shallow(<SearchResults {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render and empty container if searchResults is empty', () => {
    const wrapper = shallow(<SearchResults {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the spinner if loading', () => {
    const props = {
      ...defaultProps,
      loading: true
    };

    const wrapper = mount(<SearchResults {...props} />);

    expect(wrapper.find('svg')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger the selectProduct function when clicked', () => {
    const props = {
      ...defaultProps,
      searchResults: [
        {
          title: 'Some title',
          raw: {
            tpprixnormal: 1.5,
            tpproductid: '123',
            tpformat: '330ml',
            tpcategorie: 'beer'
          },
          percentScore: 12.34
        }
      ]
    };
    const wrapper = shallow(<SearchResults {...props} />);
    wrapper.find('Link').simulate('click');

    expect(props.selectProduct).toHaveBeenCalled();
    expect(props.selectProduct).toHaveBeenCalledWith({
      percentScore: 12.34,
      raw: {
        tpcategorie: 'beer',
        tpformat: '330ml',
        tpprixnormal: 1.5,
        tpproductid: '123'
      },
      title: 'Some title'
    });
  });
});
