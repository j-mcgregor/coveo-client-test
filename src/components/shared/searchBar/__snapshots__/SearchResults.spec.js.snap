// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should render and empty container if searchResults is empty 1`] = `ShallowWrapper {}`;

exports[`SearchResults should render and the items if present 1`] = `ShallowWrapper {}`;

exports[`SearchResults should render the spinner if loading 1`] = `ReactWrapper {}`;
